pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -531,7 +539,17 @@ public boolean commitSegment(String rawTableName, final String committingSegment
 
     // TODO Introduce a controller failure here for integration testing
 
-    updateHelixIdealState(realtimeTableName, newInstances, committingSegmentNameStr, newSegmentNameStr);
+    // When multiple segments of the same table complete around the same time it is possible that
+    // the idealstate udpate fails due to contention. We serialize the updates to the idealstate
+    // to reduce this contention. We may still contend with RetentionManager, or other updates
```

Comment:
+1 on batching. Helix applies this logic internally on updating current states but does not expose any API to clients. @jfim implemented something similar on handling notifications from Helix. It will be good to have a generic API to batch notifications and writes to any HelixEntity. This can be done in another PR. Having a lock per table will eliminate failures due to race conditions on the client side. The only failure will be because of the race condition on ZK which won't happen since all updates to IS are done from leader controller.